SQL: Add records to a db

sulthan-ahmed opened the pull request:
There are some Issues with the PR

* I'm not sure what to test

* if sql server stops and the user submits the form, the app crashes.
Ideally, it would be good for the app provided a friendly message to try again

~* The app does not attempt to reconnect when the sql server restarts~

sulthan-ahmed committed:
Post data to sql db

- Connect to a db on start up
- reuse the db connection when posting data to db
- deal with scenarios where a user does not enter data for research_date & accessneeds.  This will automatically add nulls
- bug: mysql needs single quotes for inserting values into a db even though it is a string

sulthan-ahmed committed:
Update script for linting

Also sort out linting issues

sulthan-ahmed committed:
Amend Save Data behaviour to sql

Reformat the code so the data to the db and then super is called

daniel-ac-martin commented on config.js:
```diff
@@ -0,0 +1,12 @@
+
+/* eslint no-process-env: 0 */
+module.exports = {
+  sql: {
```
As you only currently support mysql this may as well be `mysql` rather than `sql`.

sulthan-ahmed replied:
fair

daniel-ac-martin commented on config.js:
```diff
@@ -0,0 +1,12 @@
+    host: process.env.SQL_HOST,
+    user: process.env.SQL_USER,
+    password: process.env.SQL_PASSWORD,
+    db: 'digitalInclusion',
```
Set this via environment variable.

daniel-ac-martin commented on config.js:
```diff
@@ -0,0 +1,12 @@
+    user: process.env.SQL_USER,
+    password: process.env.SQL_PASSWORD,
+    db: 'digitalInclusion',
+    table: 'research_data'
```
This could probably just be a const elsewhere in the app. (Typically just in one file.)

JoeChapman replied:
also, this is in config, but you don't refer to it anywhere?
You use the literal string `research_data` in the sql statement.

sulthan-ahmed replied:
yeah good point @JoeChapman

daniel-ac-martin commented on apps/digital-inclusion/behaviours/save-data.js:
```diff
@@ -0,0 +1,55 @@
+    // build sql statement to deal with empty researchDate or/and accessNeeds,
+    // this will mean a null will go into the db
+    // bug: mysql needs insert data to be wrapped in quotations '' even though it is a string
+    let sql = `INSERT INTO research_data SET project='${project}', category='${category}'`;
```
I would farm this work out into its own module. (Probably you would call that the 'model' or something.)

That way it will be easier to test as you have less of a dance to do with instantiating the ES6 class.

daniel-ac-martin replied:
Your 'model' module would probably offer CRUD (Create, Read, Update & Delete) functionality plus maybe 'Search'. (Though don't bother implementing anything you don't currently need.)

So the code here would just do something like:
```
const model = require('some/path/model/digital-inclusion');

module.exports = superclass => class extends superclass {
  saveValues(req, res, next) {
    model.create({
      project: req.values.project,
      category: req.values.category
    });
  }
};
```

daniel-ac-martin replied:
That's how I would do it anyway.

Now on to the actual SQL that you have here...

The way that you are putting the form data into the SQL is _extremely_ dangerous. Read up on 'SQL injection' to see what I mean. (Give me a shout if it doesn't make sense.)

daniel-ac-martin replied:
I should add that modern libraries generally provide a way of safely doing this out of the box.

sulthan-ahmed committed:
Amend config

use env variable for db name
call table a const in the save data sql behaviour

sulthan-ahmed commented:
Thanks @daniel-ac-martin very helpful comments.  I'll make the changes I can.  I was aware of the danger of sql injection but wasn't quite sure how to ensure my code isn't susceptible to it.

Also, have you suggestions about dealing with sql server when it's down? it errors then it tries to connect here https://github.com/UKHomeOffice/digital-inclusion/pull/6/files#diff-3b4def45c21f84bed1c6514082004b05R36
but then it tries to run the query where obviously the connection has failed here
https://github.com/UKHomeOffice/digital-inclusion/pull/6/files#diff-3b4def45c21f84bed1c6514082004b05R42

I'm thinking of rendering a server exception page, exit and not go to next function.  

I'll do with this after I've addressed your comments

daniel-ac-martin commented:
@sulthan-ahmed: It's not ideal but a simple solution would be to simply shut-down if you can't connect to MySQL. Or to put it another way only listen to for HTTP requests once you've connected to the database.

JoeChapman commented:
Mysql crashing app, connection pools etc https://github.com/mysqljs/mysql/issues/1167

sulthan-ahmed commented:
@daniel-ac-martin that does not sound pleasant to the user at all.

thanks @JoeChapman I'll take a look

daniel-ac-martin commented:
@sulthan-ahmed: How often you do expect the database to go down? Remember that kubernetes will restart your app when it crashes. I don't think the UX would be intolerable.

sulthan-ahmed commented:
good question @daniel-ac-martin 
I guess I wouldn't too often and you'd have much more experience than me.

Also, I'd love for this to be on ACP and for it to be using kubernetes but I'm not sure if that will be the case at the moment

daniel-ac-martin commented:
If the platform can't auto-heal it then the app shouldn't crash just because the DB is down. Probably the DB pool library can handle the DB going down. Worth reading up on to work out exactly what will happen. (Or better test it and see.)

sulthan-ahmed commented:
that makes sense @daniel-ac-martin

sulthan-ahmed committed:
Move saving to sql function to a model function

It's easier to manage the functionality if it is separated out

sulthan-ahmed committed:
Bugfix: get correct save model parameters

The wrong attributes were being called for the options object in the create function for the save model

sulthan-ahmed committed:
Amend SQL statement so it prevents SQL injection

sulthan-ahmed commented:
@daniel-ac-martin I've made the changes you've requested.  

Are you happy that the app is now protected from SQL injections?

I appreciate there are no tests but I don't want to invest in writing a unit test if I'm not implementing the save model properly.

daniel-ac-martin commented on apps/digital-inclusion/behaviours/save-data.js:
```diff
@@ -1,55 +1,25 @@
+//   database: config.mysql.db
+// });
 
-const table = 'research_data';
```
Delete rather than comment out. You can always look at old commits to see the old code.

sulthan-ahmed replied:
I agree, I just forgot to delete it.  I'll do that now

daniel-ac-martin commented on apps/digital-inclusion/models/save.js:
```diff
@@ -0,0 +1,53 @@
+      console.error('error connecting: ' + err.stack);
+    } else {
+      // eslint-disable-next-line no-console
+      console.info('reconnected to the db!');
```
Have you re-connected?

sulthan-ahmed replied:
this is the console log left over from before, I'll change to 
'connect to db or use existing db connection'
Or is that too much information

daniel-ac-martin commented on apps/digital-inclusion/models/save.js:
```diff
@@ -0,0 +1,53 @@
+    } else {
+      // eslint-disable-next-line no-console
+      console.info('reconnected to the db!');
+        connection.query(sql, [project, category, researchDate, accessNeeds], (error, result) => {
```
Why the extra indentation?

daniel-ac-martin commented:
Yes, I think the SQL injection vulnerability is gone. Made a few minor comments though.

daniel-ac-martin commented on app.js:
```diff
@@ -1,3 +1,22 @@
+
+const mysql = require('mysql');
+const config = require('./config');
+const connection = mysql.createConnection({
```
Is this correct?

You are using create pool elsewhere. Surely you are only using one?

sulthan-ahmed replied:
You're right, I didn't need this. I just checked.  I'll remove it

daniel-ac-martin commented on apps/digital-inclusion/models/save.js:
```diff
@@ -0,0 +1,53 @@
+const mysql = require('mysql');
+const config = require('../../../config');
+
+const pool = mysql.createPool({
```
I'm not familiar with best practice for this library but I would suspect that this should be farmed out to a `mysql.js` or `db.js` that exports the pool.

That way you could re-use the pool across models / tables.

sulthan-ahmed replied:
what you suspect is completely logically.  However, the examples they provide in their docs are a bit more strange.  It seems, I think and I'm not sure.  You create a Pool, then you get a pool by doing getConnection.  If a pool has not been created, mysql creates one.  Otherwise it just reuses one if it is already there.

See documentation here https://github.com/mysqljs/mysql#pooling-connections

So I think what I have is fine but I'm happy to change it if it isn't correct

daniel-ac-martin replied:
I think what they are demonstrating is that you can share a single connection for multiple consecutive queries for speed. If you are only making one query at a time then you could probably follow the "Use pool directly" style.

You should still be able to move the pool into its own file. (Which would export the pool.) `save.js` could then just require in the pool.

sulthan-ahmed replied:
I tried to do what you suggested but for some reason I get `cannot read property 'getconnection' of undefined`.  For now, I'll just merge.  I'll investigate this in the future.

Thanks again

sulthan-ahmed committed:
Delete commented code in save-data.js

sulthan-ahmed committed:
Remove connection to mysql at start of app

This is not needed because sql pools are being used

sulthan-ahmed committed:
Fix save-data indentation & amend console log msg
